src/main/java/mate.academy/lib/Injector.java

```diff
@@ -1,13 +1,82 @@
 package mate.academy.lib;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+import mate.academy.service.FileReaderService;
+import mate.academy.service.ProductParser;
+import mate.academy.service.ProductService;
+import mate.academy.service.impl.FileReaderServiceImpl;
+import mate.academy.service.impl.ProductParserImpl;
+import mate.academy.service.impl.ProductServiceImpl;
+
 public class Injector {
     private static final Injector injector = new Injector();
+    private final Map<Class<?>, Object> instances = new HashMap<>();
 
     public static Injector getInjector() {
         return injector;
     }
 
     public Object getInstance(Class<?> interfaceClazz) {
-        return null;
+        Object clazzImplementationInstance = null;
+        Class<?> clazz = findImplementation(interfaceClazz);
+        checkIfAnnotationComponentPresent(clazz);
+        for (Field field : clazz.getDeclaredFields()) {
+            if (field.isAnnotationPresent(Inject.class)) {
+                clazzImplementationInstance = createNewInstance(clazz);
+                initializeField(clazzImplementationInstance, clazz, field);
+            }
+        }
+        if (clazzImplementationInstance == null) {
+            clazzImplementationInstance = createNewInstance(clazz);
+        }
+        return clazzImplementationInstance;
+    }
+
+    private void initializeField(Object clazzInstance, Class<?> clazz, Field field) {
+        Object fieldInstance = getInstance(field.getType());
+        try {
+            field.setAccessible(true);
+            field.set(clazzInstance, fieldInstance);
+        } catch (IllegalAccessException e) {
+            throw new RuntimeException("Can't initialize field value. "
+                    + "Class:" + clazz.getName()
+                    + " Field: " + field.getName());
+        }
+    }
+
+    private void checkIfAnnotationComponentPresent(Class<?> clazz) {
+        if (!clazz.isAnnotationPresent(Component.class)) {
+            throw new RuntimeException("Can't create component of " + clazz.getName() + " class.");
+        }
+    }
+
+    private Object createNewInstance(Class<?> clazz) {
+        if (instances.containsKey(clazz)) {
+            return instances.get(clazz);
+        }
+        try {
+            Constructor<?> constructor = clazz.getConstructor();
+            Object instance = constructor.newInstance();
+            instances.put(clazz, instance);
+            return instance;
+        } catch (NoSuchMethodException | InstantiationException
```

Comment:
can we combine these exceptions?

Reply:
fixed
combine this exceptions by using their parent Exception class ReflectiveOperationException